go-controller/pkg/libovsdbops/switch.go

```diff
@@ -3,338 +3,407 @@ package libovsdbops
 import (
 	"context"
 	"fmt"
-	"strings"
 
 	libovsdbclient "github.com/ovn-org/libovsdb/client"
-	"github.com/ovn-org/libovsdb/model"
 	libovsdb "github.com/ovn-org/libovsdb/ovsdb"
 
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
 )
 
-// findSwitchUUID looks up the switch in the cache and sets the UUID
-func findSwitchUUID(nbClient libovsdbclient.Client, lswitch *nbdb.LogicalSwitch) error {
-	if lswitch.UUID != "" && !IsNamedUUID(lswitch.UUID) {
-		return nil
-	}
+// LOGICAL_SWITCH OPs
+
+type switchPredicate func(*nbdb.LogicalSwitch) bool
 
+// FindLogicalSwitchesWithPredicate looks up logical swiotches from the cache
+// based on a given predicate
+func FindLogicalSwitchesWithPredicate(nbClient libovsdbclient.Client, p switchPredicate) ([]*nbdb.LogicalSwitch, error) {
+	found := []*nbdb.LogicalSwitch{}
 	ctx, cancel := context.WithTimeout(context.Background(), types.OVSDBTimeout)
 	defer cancel()
-	switches := []nbdb.LogicalSwitch{}
-	err := nbClient.WhereCache(func(item *nbdb.LogicalSwitch) bool {
-		return item.Name == lswitch.Name
-	}).List(ctx, &switches)
-	if err != nil {
-		return fmt.Errorf("can't find switch %+v: %v", *lswitch, err)
-	}
+	err := nbClient.WhereCache(p).List(ctx, &found)
+	return found, err
+}
 
-	if len(switches) > 1 {
-		return fmt.Errorf("unexpectedly found multiple switches: %+v", switches)
+// GetLogicalSwitch looks up a logical switch from the cache
+func GetLogicalSwitch(nbClient libovsdbclient.Client, sw *nbdb.LogicalSwitch) (*nbdb.LogicalSwitch, error) {
+	found := []*nbdb.LogicalSwitch{}
+	opModel := OperationModel{
+		Model:          sw,
+		ModelPredicate: func(item *nbdb.LogicalSwitch) bool { return item.Name == sw.Name },
+		ExistingResult: &found,
+		OnModelUpdates: nil, // no update
+		ErrNotFound:    true,
+		BulkOp:         false,
 	}
 
-	if len(switches) == 0 {
-		return libovsdbclient.ErrNotFound
+	m := NewModelClient(nbClient)
+	_, err := m.CreateOrUpdate(opModel)
+	if err != nil {
+		return nil, err
 	}
 
-	lswitch.UUID = switches[0].UUID
-	return nil
+	return found[0], nil
 }
 
-// findSwitches returns all the current logicalSwitches
-func findSwitches(nbClient libovsdbclient.Client) ([]nbdb.LogicalSwitch, error) {
-	switches := []nbdb.LogicalSwitch{}
-	ctx, cancel := context.WithTimeout(context.Background(), types.OVSDBTimeout)
-	defer cancel()
-	err := nbClient.List(ctx, &switches)
-	if err != nil {
-		return nil, fmt.Errorf("can't find Locial Switches err: %v", err)
+// CreateOrUpdateLogicalRouter creates or updates the provided logical switch
+func CreateOrUpdateLogicalSwitch(nbClient libovsdbclient.Client, sw *nbdb.LogicalSwitch) error {
+	opModel := OperationModel{
+		Name:           sw.Name,
+		Model:          sw,
+		ModelPredicate: func(item *nbdb.LogicalSwitch) bool { return item.Name == sw.Name },
```

Comment:
this goes for all the times you are looking for a switch by the name. I know that ovn-kubernetes gives the switches a unique name but the ovn nbdb does not guarantee that switches won't share the same names. Should we be ensuring that only the switch made by ovn-kube is being edited?

Reply:
Hmm Well relying on the fact that ovn-kubernetes gives the switches a unique name is not something that has been introduced in this PR, we were already doing so before. Only ovnk should be adding switches to NBDB.
And I can't come up with a good way to do that, how do we distinguish a switch added from ovnk from other switch added by a third party if the third party is replicating what ovnk does?

Reply:
@jcaamano we have a cache of switches the LogicalSwitchManager has `GetUUID()` which can get the UUID's of logical switches by name